pkg/gatewaysecret/handler.go

```diff
@@ -0,0 +1,137 @@
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/lifecycle-manager/internal/pkg/flags"
+)
+
+const (
+	GatewaySecretName = "istio-gateway-secret"
+	kcpRootSecretName = "klm-watcher"
+	kcpCACertName     = "klm-watcher-serving"
+	istioNamespace    = flags.DefaultIstioNamespace
+)
+
+type GatewaySecretHandler struct {
+	*secretManager
+}
+
+func NewGatewaySecretHandler(kcpClient client.Client) *GatewaySecretHandler {
+	return &GatewaySecretHandler{
+		secretManager: &secretManager{
+			kcpClient: kcpClient,
+		},
+	}
+}
+
+func (gsh *GatewaySecretHandler) ManageGatewaySecret(rootSecret *apicorev1.Secret) error {
+	gwSecret, err := gsh.findGatewaySecret()
+
+	if isNotFound(err) {
+		return gsh.handleNonExisting(rootSecret)
+	}
+	if err != nil {
+		return err
+	}
+
+	return gsh.handleExisting(rootSecret, gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleNonExisting(rootSecret *apicorev1.Secret) error {
+	gwSecret := gsh.newGatewaySecret(rootSecret)
+	return gsh.create(context.Background(), gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleExisting(rootSecret *apicorev1.Secret, gwSecret *apicorev1.Secret) error {
+	caCert := certmanagerv1.Certificate{}
+	if err := gsh.kcpClient.Get(context.Background(),
+		client.ObjectKey{Namespace: istioNamespace, Name: kcpCACertName},
+		&caCert); err != nil {
+		return fmt.Errorf("failed to get CA certificate: %w", err)
+	}
+
+	if gwSecretLastModifiedAtValue, ok := gwSecret.Annotations[LastModifiedAtAnnotation]; ok {
+		if gwSecretLastModifiedAt, err := time.Parse(time.RFC3339, gwSecretLastModifiedAtValue); err == nil {
+			if caCert.Status.NotBefore != nil && gwSecretLastModifiedAt.After(caCert.Status.NotBefore.Time) {
+				return nil
+			}
+		}
+	}
+
+	gwSecret.Data["tls.crt"] = rootSecret.Data["tls.crt"]
+	gwSecret.Data["tls.key"] = rootSecret.Data["tls.key"]
+	gwSecret.Data["ca.crt"] = rootSecret.Data["ca.crt"]
+	return gsh.update(context.Background(), gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) findGatewaySecret() (*apicorev1.Secret, error) {
+	return gsh.findSecret(context.Background(), client.ObjectKey{
+		Name:      GatewaySecretName,
+		Namespace: istioNamespace,
+	})
+}
+
+func (gsh *GatewaySecretHandler) newGatewaySecret(rootSecret *apicorev1.Secret) *apicorev1.Secret {
+	gwSecret := &apicorev1.Secret{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: apicorev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name:      GatewaySecretName,
+			Namespace: istioNamespace,
+		},
+		Data: map[string][]byte{
+			"tls.crt": rootSecret.Data["tls.crt"],
+			"tls.key": rootSecret.Data["tls.key"],
+			"ca.crt":  rootSecret.Data["ca.crt"],
+		},
+	}
+	return gwSecret
+}
+
+func WatchChangesOnRootCertificate(clientset *kubernetes.Clientset, gatewaySecretHandler *GatewaySecretHandler,
```

Comment:
I expect the scenario you designed for this function fully tested in an dedicated e2e test, for example, the root cert get updated, test your watched one should get new secret synced, right?

Reply:
but that is already being tested in the ca rotation test. The root cert is set to rotate every 1 minute and we check if the skr cert is rotated. so we are are also checking the rotation of the gateway by proxy. I would suggest that we explicitly check the gateway secret in same test, rather than a dedicated test. WDZT? @ruanxin